### ISSUE TITLE
Strict effects can break focus restoration

### ISSUE THREAD

**Issue by devongovett** [facebook/react#25979]
This isn't necessarily a bug in React's implementation, but a symptom caused by strict effects. It is pretty common to need to restore focus to another element when a component unmounts. For example, a dialog component would need to restore focus to the element that opened it for proper keyboard accessibility.

When StrictMode is enabled, components are "mounted" twice such that their effects and cleanup functions are executed, but the component isn't actually removed from the DOM in this process. If a component moves focus in its effect cleanup function, this has user visible consequences and focus won't be restored back when the component "mounts" again. Here is a simple example:

```js
function Inner({ restoreRef }) {
  useLayoutEffect(() => {
    let el = restoreRef.current;
    return () => {
      // Restore focus on unmount.
      el.focus();
    };
  }, [restoreRef]);

  return <input autoFocus />;
}
```

In this example, the browser moves focus to the input when Inner is committed to the DOM. Then, the layout effect cleanup is run, restoring focus to the provided ref. But the component hasn't actually unmounted, so focus should ideally stay on the input.

One could try to work around this by not using `autoFocus` and instead using a ref to the input and calling `focus()` on it in the `useLayoutEffect`. That way, when the component "mounts" again, the layout effect would run a second time and focus would come back to the input. However, this might have accessibility implications if focus is bouncing around rapidly. It also wouldn't work if you didn't want to auto focus an element on initial mount but still wanted focus restoration on unmount. Ideally there would be a way to avoid strict effects entirely for this use case.

React version: 18.2.0

## Steps To Reproduce

1. https://codesandbox.io/s/async-dust-k8g9uy?file=/src/App.js
2. Press the "Toggle inner" button
3. See that focus is not on the auto focused input.

## The current behavior

Strict effects with side effects in cleanup functions that cannot be undone have no way to only run when the component is actually being removed from the DOM.

## The expected behavior

Ideally some way to know if the component is actually being removed from the DOM so this can only happen then. Basically an opt-out of strict effects for this specific use case.

**Comment by sebmarkbage** [comment]
To see this behavior in production today, in React 18, you can try resuspending a Suspense boundary without `startTransition`. You're not typically supposed to do this because they should use `startTransition` but it can happen.

```
<Suspense fallback={...}>
  <ThingWithFocus />
  <SuspendTemporarily />
</Suspense>
```

If this has already mounted once, so it's visible and you have the focus, then SuspendTemporarily resuspends it'll hide the current tree but it won't delete it from the DOM. It's the same DOM nodes so that they preserve their state. Although maybe we need to put it in "Inert" mode. It's virtually gone and there should be no way to interact with it.

While the DOM nodes are technically still there, they're in an inert state and it doesn't make sense to navigate around them. So it makes sense that focus would need to go elsewhere while the boundary is loading. However, if it quickly restores, then maybe it makes sense to treat focus as if it didn't suspend.

The question isn't about what happens in DEV mode and StrictMode but what should happen in this scenario in production.

There are more cases like this we'll expand this concept to. There's a bunch of work on "Offscreen" that you can follow in the repo. There are many arguments for and against why effects should or shouldn't unmount when this happens. It's possible that we're wrong, however, we're a number of years into this and haven't found a case yet where it doesn't make sense.

**Comment by devongovett** [comment]
Yeah I think in general it should work like it unmounted (eg unsubscribing from data stores, etc.), but the focus case is different. In any case where the DOM node becomes hidden (either because it was removed from the DOM or became inert), focus should move somewhere else. The problem is that there isn't a way to distinguish these cases currently. I am not sure the right way to do that API wise (maybe a new type of effect 😬) but it does seem like a worthwhile problem to solve.

**Comment by sebmarkbage** [comment]
I'm not sure which case would need to be different - like if there was a new type of effect. Like if we removed from the document and then reinserted the same instances (to preserve state of the instance) instead of hiding it / inert, would it make a difference in this case? Focus is temporarily lost regardless.

**Comment by sebmarkbage** [comment]
Another interesting scenario is a tab switcher. If I'm switching between tabs using keyboard shortcuts. When I switch to a new tab, focus would end up somewhere inside that tab - e.g. to read a header. Then I can change focus within the tab. Then I switch to another tab, and back again. I'd say that you'd probably want to restore focus to where it was before I switched to the other tab instead of resetting to the default.

**Comment by poeschko** [comment]
Maybe this is a case where React hooks are not the right fit? Restoring focus is not the "cleanup" of an actual effect, after all. If what ultimately matters is DOM mutations, maybe a `MutationObserver` is the right answer?

**Comment by sebmarkbage** [comment]
> Maybe this is a case where React hooks are not the right fit? Restoring focus is not the "cleanup" of an actual effect, after all. If what ultimately matters is DOM mutations, maybe a MutationObserver is the right answer?

Thinking about this in terms of "real DOM" is the wrong mental model. It's an implementation detail if it's still in the DOM or not. What matters is that it's in accessible one way or another. We could also remove the DOM from the document.

There similarly is no "real" React unmount in this concept. Think of it as removing the DOM node from the document and then reinserting it later in plain imperative APIs without React. If you never reinsert it, there's no extra event that happens other than a WeakRef when it gets GC:ed.

So a component needs to be resilient to being temporarily lifted out (unmounted) and restored with the same state soon after.

My argument is that a component that restores focus to some previous owner after being unmounted, needs to also reclaim focus if it gets remounted - as long as it's still the "owner" of that focus. Otherwise that component isn't resilient to being temporarily in a suspended state or deeply nested inside a tab switcher like that.

**Comment by shreya-as** [comment]
_Maybe this is a case where React hooks are **not** the right fit_

**Comment by frfancha** [comment]
Don't use autofocus.
Give focus to the component input in a useEffect on a ref to it.
And restore focus to the previously focused item in the clean up of that useEffect.
That way everything will work well including in strict mode.
As soon as you "clean" things in a useEffect that you haven't set in that same useEffect, like in your example you're trying to "clean" the autofocus, you will have issues in some scenarios.
These issues might be reduced by removing strict mode, but not completely. Strict mode is there to help you to detect what will go wrong one day so better to keep it, and write proper clean code (cleaning what was set in the useEffect and not something set externally)

**Comment by devongovett** [comment]
Auto focus is not the only case where this can happen, as I covered in the original issue.

**Comment by frfancha** [comment]
> Auto focus is not the only case where this can happen, as I covered in the original issue.

What is "this" ?

If you mean "the component is mounted twice but not removed from the DOM" that is as designed.
React will one day optimize performance trying to minimize the new DOM object creation.
Imagine you're writing a tab component with one tab visible at a time.
Right now if you just render your visible tab conditionally a new component must be created anytime the user changes the visible tab. The current workaround is to use display:none instead of removing components from the virtual DOM and therefore from the DOM.
That won't be necessary anymore. But to be prepared for that logic you must stop thinking mounted = created in the DOM. That's what the strict mode is for: if your code is written correctly for strict mode, it will be correct when that optimization will be in production.

And therefore the general conclusion stays the same: if you need to do something when the component is "mounted", the correct way to do it is a useEffect on [] with a clean action in that same useEffect.

> It also wouldn't work if you didn't want to auto focus an element on initial mount but still wanted focus restoration on unmount.

No that will work perfectly.

> Ideally there would be a way to avoid strict effects entirely for this use case.

Bad idea, writing code that works only without strict mode is writing code that will break in the future.
But if you really want no strict mode, that is a dev parameter only, once your code is built it disappear. And to remove it during dev just put Fragment instead of StrictMode in your index file.

**Comment by devongovett** [comment]
I don't appreciate being talked down to. I know what strict mode is any why it's useful. But as discussed in the comments above, I think in this case it is not necessarily simulating what _should_ happen in a real world case (like the tabs example).

Here is a very simple example showing the difference without auto focus. Focus should be moved to the outer input element on unmount. In the strict mode version, it is also moved there on mount. https://codesandbox.io/s/cranky-http-zg28jv?file=/src/App.js

In this case the mount effect and unmount effect are not opposites of each other (nor should they be), so strict mode becomes observable to the user.

**Comment by frfancha** [comment]
Interesting demo.
For me the no strict mode version doesn't restore the focus when hidden unless I replace useLayoutEffect by useEffect, is it the same at your side?
And indeed the strict mode version restore the focus immediately as expected in strict mode.
It won't happen once built though, so I would say strict mode is observable to the developer more than the user.

If I adapt your useEffect in Inner, it works as expected with or without StrictMode:
```
  const focusRef = useRef();
  useEffect(() => {
    if (focusRef.current) {
      focusRef.current.focus();
    }
    let el = restoreRef.current;
    return () => {
      focusRef.current = document.activeElement;
      // Restore focus to the outer input on unmount.
      el.focus();
    };
  }, [restoreRef]);

```

What do you think?

**Comment by devongovett** [comment]
Yeah it seems like it was not working in Chrome because the browser moved focus to the button instead. Adding `preventDefault` to `onMouseDown` made it behave consistently cross browser.

Your approach of storing the active element on unmount and restoring on mount would work around the strict mode case, but it could also break other cases. For example, imagine the tabs case with the future React state restoration feature. The component could be re-mounted some time later, at which point it may or may not be desired to restore focus to the previously focused element from when it last unmounted. Focus could have moved somewhere completely different by that point and it might not make sense to restore it anymore. There may not be an obvious way to distinguish between this real-world case and the strict mode case. That's why I'm suggesting that React add a way to do so.

**Comment by frfancha** [comment]
Yes you're right my code in no-strict mode is at best useless, and often buggy.

In fact StrictMode does all it can to make fake unmounts from StrictMode indistinguishable from "true" unmounts.
Let's even forget about focus, imagine you need to maintain a counter on the 'true' number of unmounts there is just no way to do it in the unmounted component itself. You can do it in the component using the unmounted one though.

So... options:

A-ignore the fact that the behaviour is not exactly the one you want in strict mode, strict mode disappears once your code is built anyway? (Not super I know, especially if that code is reused by other teams)

B-decide that the correct way to do something on unmount of Inner is not to pass a ref to Inner for Inner to do it itself, but to do it yourself in the component using Inner, in a useEffect on the condition displaying Inner

C-consider that if it makes sense to put the focus on X when Inner in unmounted, it means also that it makes sense for Inner to capture the focus when it appears, which restores the symmetry of your mount/unmount operations and solve the issue.

D-develop in non Strict mode

I would go for C or B, but it isn't my code of course, so that's easy to say for me.

**Comment by frfancha** [comment]
Not super useful, but still here the code answering to the original scenario using autofocus:
```
function Inner({ restoreRef }) {
  const myInputRef = useRef();
  useEffect(() => {
    myInputRef.current.focus();
    return () => {
      // Restore focus to the outer input on unmount.
      restoreRef.current.focus();
    };
  }, []);

  return (
    <div style={{ padding: 10, background: "pink" }}>
      <input ref={myInputRef} />
    </div>
  );
}
```
